### DISCUSSION TITLE
Changing TransformerParameters - input requested!

### DISCUSSION THREAD

**Discussion by winedarksea** [winedarksea/AutoTS#27] [Ideas]
I have been working on completely replacing the format of TransformationParameters. It is a very subjective process, so any ideas for how you would do it are appreciated!
The new JSON parameters as I see it now, would look something like this:
```
{
 "fillna": "mean",
 "transformations": {0: "SomeTransformer", 1: "AnotherTransformer", 2: "FancyTransformer"}, 
 "transformation_params":  {0: null, 1: 25, 2: {"param1": 10, "param2": "fancy"}}, 
}
```
A `transformer_max_depth` parameter will be added adjusting how many transformations will be allowed. Also new, a `transformer_list` allowing users to customize what transformers they allow, probably with "fast" and "all" aliases.

All of the previously separated tasks - outlier removal and discretization, for example, are now just another type of Transformer. These new Transformers can be a bit confusing since their `inverse_transform` methods don't actually undo the transformation - although since most people won't be using those lower level functions anyway, that's probably a minor issue.

New:
- Much simpler and more aesthetic
- Much more flexible

Old:
- enforced a 'standard' transformer pattern that most data scientists loosely follow - remove outliers, fillna, do a mathematical transformation, detrend the series, and then, sometimes, decompose the seasonality - all in more or less the same order. 
- easier to customize probabilities for model ordering in randomly generated parameter sets.

I think the new layout could be a bit slower to optimize, simply because, being free-form, it can generate so many more possible options - many of which would be insignificant changes, or just outright bad. On the other hand, it should in theory also allow superior, more accurate, transformation combinations. Setting a transformer_max_depth=2, say, would be one way to enforce simple transformations and speed up optimization. In general, users would have more control over transformations with the new layout.

Here is the current (Old) TransformationParameters layout for reference:
```
{'outlier_method': 'clip',
 'outlier_threshold': 3,
 'outlier_position': 'first',
 'fillna': 'ffill',
 'transformation': 'RobustScaler',
 'second_transformation': None,
 'transformation_param': None,
 'detrend': None,
 'third_transformation': 'MinMaxScaler',
 'transformation_param2': None,
 'fourth_transformation': None,
 'discretization': 'center',
 'n_bins': 10,
 'grouping': None,
 'reconciliation': None,
 'coerce_integer': False,
 'context_slicer': None}
```

**Comment by winedarksea** [comment]
@mreismendes care to share your thoughts on this as well?

**Comment by mreismendes** [comment]
@winedarksea I always tend to prefer things that are intuitive and offer great usability. It seems that your purpose goes in that way. Once the idea is to generate more combinations of parameters, it sounds good.
